python/GafferArnoldTest/ArnoldRenderTest.py

```diff
@@ -1181,6 +1181,69 @@ def testDefaultLightsMistakesDontForceLinking( self ) :
 			self.assertEqual( arnold.AiArrayGetNumElements( arnold.AiNodeGetArray( sphere, "light_group" ) ), 0 )
 			self.assertFalse( arnold.AiNodeGetBool( sphere, "use_light_group" ) )
 
+	def testLightLinkingWarnings( self ) :
+
+		# Emulate a meshlight that has been set up sloppily - it is filtered to 4 locations, some actually
+		# have meshes, some don't
+		lightSphere = GafferScene.Sphere()
+		lightInvalid = GafferScene.Group()
+
+		lightGroup = GafferScene.Group()
+		lightGroup["name"].setValue( "lightGroup" )
+		lightGroup["in"][0].setInput( lightSphere["out"] ) # Has a mesh
+		lightGroup["in"][1].setInput( lightSphere["out"] ) # Has a mesh
+		lightGroup["in"][2].setInput( lightInvalid["out"] ) # Doesn't have a mesh
+		lightGroup["in"][3].setInput( lightInvalid["out"] ) # Doesn't have a mesh
+
+		meshLightFilter = GafferScene.PathFilter()
+		meshLightFilter["paths"].setValue( IECore.StringVectorData( [ "/lightGroup/*" ] ) )
+
+		meshLight = GafferArnold.ArnoldMeshLight()
+		meshLight["in"].setInput( lightGroup["out"] )
+		meshLight["filter"].setInput( meshLightFilter["out"] )
+
+		geoSphere = GafferScene.Sphere()
+		geoGroup = GafferScene.Group()
+		geoGroup["name"].setValue( "geoGroup" )
+		for i in range( 20 ):
+			geoGroup["in"][i].setInput( geoSphere["out"] )
+
+		group = GafferScene.Group()
+		group["in"][0].setInput( geoGroup["out"] )
+		group["in"][1].setInput( meshLight["out"] )
+
+		attributeFilter = GafferScene.PathFilter()
+		attributeFilter["paths"].setValue( IECore.StringVectorData( [ "/group/geoGroup/*" ] ) )
+
+		attributes = GafferScene.StandardAttributes()
+		attributes["in"].setInput( group["out"] )
+		attributes["filter"].setInput( attributeFilter["out"] )
+		attributes["attributes"]["linkedLights"]["enabled"].setValue( True )
+		# Link some ( but not all ) lights, so we have to do actual light linking
+		attributes["attributes"]["linkedLights"]["value"].setValue(
+			"/group/lightGroup/sphere1 /group/lightGroup/group /group/lightGroup/group1"
+		)
+
+		render = GafferArnold.ArnoldRender()
+		render["in"].setInput( attributes["out"] )
+		render["mode"].setValue( render.Mode.SceneDescriptionMode )
+		render["fileName"].setValue( self.temporaryDirectory() + "/test.ass" )
+
+		# Don't really understand why a regular `with CapturingMessageHandler` doesn't work here
```

Comment:
I _think_ this is because the current message handler is per-thread, and `execute()` will do work on multiple threads. It is possible to pass a message handler to the renderer constructor which it is then responsible for using on all threads, although that is not useful here because the renderer is being created by the node. But perhaps when no message handler is passed, we should store the current handler from the constructing thread?

More concretely, I'm wondering if we should do `m_messageHandler( messageHandler ? messageHandler : MessageHandler::currentHandler()` [here](https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/IECoreArnoldPreview/Renderer.cpp#L3897).

Reply:
In a quick test, that fix doesn't seem sufficient in itself for testing this.  It looks like `outputLightLinks()` isn't called with the renderer's message handler scoped, so if we wanted that to work reliably for everything we would have to do something like caching the message handler in ArnoldObject, or adding an interface to IECoreScenePreview::Renderer for accessing the renderer's message handler so outputLightLinks() could get the mesage handler from the renderer and scope it.